### PR TITLE
Use legacy coordinate in metric of VOC

### PR DESCRIPTION
**Pull request by jshilong**
## Motivation
In the official [metric implementation](http://host.robots.ox.ac.uk/pascal/VOC/voc2012/VOCdevkit_18-May-2011.tar) of VOC dataset, it adopts the  Legacy coordinate system in mmdet 1.x, which means h, w should be computed as `x2 - x1 + 1` and `y2 - y1 + 1`.

## Modification
I add argument `Legacy_coordinate` to `eval_map`， `tpfp_default` and `tpfp_imagenet`，`bbox_overlaps`  to control whether to use legacy coordinate system when caculate the width and length.

## BC-breaking 
None

<!DOCTYPE html><meta http-equiv="content-type" content="text/html; charset=utf-8">

before/after | faster
-- | --
mAP | 79.5/79.6
recall | 93.4/93.5

### CODE REVIEW

**Comment by ZwwWayne** [comment]
Need unittests.

File: tests/test_metrics/test_box_overlap.py [hunk 1]
```diff
@@ -103,3 +105,34 @@ def _construct_bbox(num_bbox=None):
...
+    bboxes2, _ = _construct_bbox(num_bbox)
+    ious = recall_overlaps(bboxes1, bboxes2, 'iou', extra_length=0)
+    assert ious.shape == (num_bbox, num_bbox)
+    assert np.all(ious >= -1) and np.all(ious <= 1)
+
+    ious = recall_overlaps(bboxes1, bboxes2, 'iou', extra_length=1)
+    assert ious.shape == (num_bbox, num_bbox)
+    assert np.all(ious >= -1) and np.all(ious <= 1)
+
+    # extra_length should be in (1,2)
```
Comment (hhaAndroid) [inline tests/test_metrics/test_box_overlap.py:134]: extra_length should be in (0,1)

File: mmdet/core/evaluation/mean_ap.py [hunk 2]
```diff
@@ -164,13 +178,24 @@ def tpfp_default(det_bboxes,
             of shape (k, 4). Default: None
         iou_thr (float): IoU threshold to be considered as matched.
             Default: 0.5.
-        area_ranges (list[tuple] | None): Range of bbox areas to be evaluated,
-            in the format [(min1, max1), (min2, max2), ...]. Default: None.
+        area_ranges (list[tuple] | None): Range of bbox areas to be
+            evaluated, in the format [(min1, max1), (min2, max2), ...].
+            Default: None.
+        use_legacy_coordinate (bool): Whether to use coordinate system in
+            mmdet v1.x. which means width, height should be
```
Comment (hhaAndroid) [inline mmdet/core/evaluation/mean_ap.py:185]: Which means

File: mmdet/core/evaluation/bbox_overlaps.py [hunk 3]
```diff
@@ -1,21 +1,25 @@
...
 
-def bbox_overlaps(bboxes1, bboxes2, mode='iou', eps=1e-6):
+def bbox_overlaps(bboxes1, bboxes2, mode='iou', eps=1e-6, extra_length=0.):
     """Calculate the ious between each bbox of bboxes1 and bboxes2.
 
     Args:
-        bboxes1(ndarray): shape (n, 4)
-        bboxes2(ndarray): shape (k, 4)
-        mode(str): iou (intersection over union) or iof (intersection
+        bboxes1 (ndarray): shape (n, 4)
```
Comment (hhaAndroid) [inline mmdet/core/evaluation/bbox_overlaps.py:8]: Shape (n, 4)

File: mmdet/core/evaluation/mean_ap.py [hunk 4]
```diff
@@ -98,14 +109,15 @@ def tpfp_imagenet(det_bboxes,
...
+            det_areas = (
+                det_bboxes[:, 2] - det_bboxes[:, 0] + extra_length) * (
+                    det_bboxes[:, 3] - det_bboxes[:, 1] + extra_length)
             for i, (min_area, max_area) in enumerate(area_ranges):
                 fp[i, (det_areas >= min_area) & (det_areas < max_area)] = 1
         return tp, fp
-    ious = bbox_overlaps(det_bboxes, gt_bboxes - 1)
-    gt_w = gt_bboxes[:, 2] - gt_bboxes[:, 0]
-    gt_h = gt_bboxes[:, 3] - gt_bboxes[:, 1]
+    ious = bbox_overlaps(det_bboxes, gt_bboxes - 1, extra_length=extra_length)
```
Comment (RangiLyu) [inline mmdet/core/evaluation/mean_ap.py:118]: `bbox_overlaps` has no `extra_length` kwarg

File: mmdet/core/evaluation/recall.py [hunk 5]
```diff
@@ -97,7 +106,8 @@ def eval_recalls(gts,
         if gts[i] is None or gts[i].shape[0] == 0:
             ious = np.zeros((0, img_proposal.shape[0]), dtype=np.float32)
         else:
-            ious = bbox_overlaps(gts[i], img_proposal[:prop_num, :4])
+            ious = bbox_overlaps(
```
Comment (RangiLyu) [inline mmdet/core/evaluation/recall.py:109]: `bbox_overlaps` has `no extra_length` kwarg

File: mmdet/core/evaluation/bbox_overlaps.py [hunk 6]
```diff
@@ -1,21 +1,32 @@
...
+            Default: False.
 
     Returns:
-        ious(ndarray): shape (n, k)
+        ious (ndarray): Shape (n, k)
     """
 
     assert mode in ['iou', 'iof']
-
+    if not use_legacy_coordinate:
```
Comment (RangiLyu) [inline mmdet/core/evaluation/bbox_overlaps.py:26]: Why not add the extra_length to bbox before passing them to the bbox_overlaps function?

Comment (jshilong) [reply mmdet/core/evaluation/bbox_overlaps.py:26]: I believe that all metric-related code should strictly follow the official implementation, and equivalent modification may cause unnecessary misunderstandings.

File: tests/test_metrics/test_mean_ap.py [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+
+
+def test_tpfp_imagenet():
+
+    tpfp_imagenet(
+        det_bboxes,
+        gt_bboxes,
+        gt_bboxes_ignore=gt_ignore,
+        use_legacy_coordinate=True)
+    tpfp_imagenet(
```
Comment (RangiLyu) [inline tests/test_metrics/test_mean_ap.py:21]: It would be better to have a reference result.

**Review by AronLin — APPROVED** [review]

**Comment by jshilong** [comment]
@ZwwWayne We can not rename the `bbox_overlaps` to `voc_bbox_overlaps` and change the default behavior, because it would be used in all datasets when evaluating the recall of RPN

**Comment by ZwwWayne** [comment]
> @ZwwWayne We can not rename the `bbox_overlaps` to `voc_bbox_overlaps` and change the default behavior, because it would be used in all datasets when evaluating the recall of RPN

Then update the docstring, please.

**Comment by jshilong** [comment]
> > @ZwwWayne We can not rename the `bbox_overlaps` to `voc_bbox_overlaps` and change the default behavior, because it would be used in all datasets when evaluating the recall of RPN
> 
> Then update the docstring, please.

Done and I add more unitests for metrics

**Review by ZwwWayne — APPROVED** [review]

**Pull request merged**